File: app/src/main/java/de/westnordost/streetcomplete/quests/QuestUtil.java
package de.westnordost.streetcomplete.quests;

import android.content.res.Resources;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Html;
import android.text.Spanned;

import java.util.Collections;
import java.util.Map;

import de.westnordost.osmapi.map.data.Element;
import de.westnordost.streetcomplete.data.QuestType;
import de.westnordost.streetcomplete.data.osm.OsmElementQuestType;

public class QuestUtil
{
	public static String getTitle(
		@NonNull Resources r, @NonNull QuestType questType, @Nullable Element element)
	{
		Map<String,String> tags = getTags(element);
		return r.getString(getQuestTitleResId(questType, tags), getElementName(tags));
	}

	public static Spanned getHtmlTitle(
		@NonNull Resources r, @NonNull QuestType questType, @Nullable Element element)
	{
		Map<String,String> tags = getTags(element);
		String name = getElementName(tags);
		String spanName = name != null ? "<i>"+Html.escapeHtml(name)+"</i>" : null;

		return Html.fromHtml(r.getString(getQuestTitleResId(questType, tags), spanName));
	}

	private static Map<String,String> getTags(@Nullable Element element)
	{
		return element != null && element.getTags() != null ? element.getTags() : Collections.emptyMap();
	}

	private static int getQuestTitleResId(QuestType questType, Map<String,String> tags)
	{
		if(questType instanceof OsmElementQuestType)
		{
			return ((OsmElementQuestType) questType).getTitle(tags);
		}
		return questType.getTitle();
	}

	private static String getElementName(Map<String,String> tags)
	{
		return tags.get("name");
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/QuestUtil.kt
package de.westnordost.streetcomplete.quests

import android.content.res.Resources
import android.text.Html
import android.text.Spanned
import androidx.core.text.parseAsHtml
import de.westnordost.streetcomplete.data.osm.osmquests.OsmElementQuestType
import de.westnordost.streetcomplete.data.quest.QuestType

fun Resources.getQuestTitle(questType: QuestType, tags: Map<String, String>): String {
    val arguments = getTemplateArguments(questType, tags)
    return getString(getQuestTitleResId(questType, tags), *arguments)
}

fun Resources.getHtmlQuestTitle(questType: QuestType, tags: Map<String, String>): Spanned {
    val arguments = getTemplateArguments(questType, tags)
    val spannedArguments = arguments.map { "<i>" + Html.escapeHtml(it) + "</i>" }.toTypedArray()
    return getString(getQuestTitleResId(questType, tags), *spannedArguments).parseAsHtml()
}

private fun getTemplateArguments(questType: QuestType, tags: Map<String, String>): Array<String> =
    ((questType as? OsmElementQuestType<*>)?.getTitleArgs(tags)) ?: emptyArray()

private fun getQuestTitleResId(questType: QuestType, tags: Map<String, String>) =
    (questType as? OsmElementQuestType<*>)?.getTitle(tags) ?: questType.title
